xdsl/dialects/builtin.py

```diff
@@ -386,18 +385,17 @@ def verify(self) -> None:
     @staticmethod
     @builder
     def from_dict(data: dict[str | StringAttr, Attribute]) -> DictionaryAttr:
-        to_add_data: dict[str, Attribute] = {}
+        to_add_data = {}
         for k, v in data.items():
+            # try to coerce keys into StringAttr
+            if isinstance(k, StringAttr):
+                k = k.data
+            # if coercion fails, raise KeyError!
             if not isinstance(k, str):
-                if isinstance(k, StringAttr):
-                    to_add_data[k.data] = v
-                else:
-                    raise TypeError(
-                        f"Attribute DictionaryAttr expects keys to"
-                        f" be of type StringAttr or str, but {type(k)} provided"
-                    )
-            else:
-                to_add_data[k] = v
+                raise TypeError(
+                    f"Attribute DictionaryAttr expects keys to"
+                    f" be of type str or str, but {type(k)} provided")
```

Comment:
Be of type "str or str"?

Reply:
Oops, as I said, much confusion was had over this, I'll clean it up!